pkg/standalone-utils/contracts/BatchRelayer.sol

```diff
@@ -0,0 +1,168 @@
+        _sweepETH();
+    }
+
+    function joinAndStake(
+        bytes32 poolId,
+        address payable recipient,
+        IVault.JoinPoolRequest calldata joinPoolRequest
+    ) external payable {
+        getVault().joinPool{ value: msg.value }(poolId, msg.sender, address(this), joinPoolRequest);
+
+        IERC20 bpt = IERC20(_getPoolAddress(poolId));
+        uint256 bptAmount = bpt.balanceOf(address(this));
+
+        // If necessary, give staking contract allowance to take BPT
+        if (bpt.allowance(address(this), address(getStakingContract())) < bptAmount) {
+            bpt.approve(address(getStakingContract()), type(uint256).max);
+        }
+
+        getStakingContract().stake(bpt, bptAmount, recipient);
+        _sweepETH();
+    }
+
+    function swapAndExit(
+        bytes32 poolId,
+        address payable recipient,
+        IVault.ExitPoolRequest memory request,
+        IVault.SwapKind kind,
+        IVault.BatchSwapStep[] calldata swaps,
+        IAsset[] calldata assets,
+        int256[] calldata limits,
+        uint256 deadline
+    ) external payable {
+        // We can't output tokens to the user's internal balance
+        // as they need to have BPT on their address for the exit
+        IVault.FundManagement memory funds = IVault.FundManagement({
+            sender: msg.sender,
+            fromInternalBalance: false,
+            recipient: msg.sender,
+            toInternalBalance: false
+        });
+        int256[] memory swapAmounts = getVault().batchSwap{ value: msg.value }(
+            kind,
+            swaps,
+            assets,
+            funds,
+            limits,
+            deadline
+        );
+
+        // Prevent stack-too-deep
+        {
+            // Read amount of BPT from BatchSwap return value
+            uint256 bptAmount;
+            IAsset bpt = IAsset(_getPoolAddress(poolId));
+            for (uint256 i; i < assets.length; i++) {
+                if (assets[i] == bpt) {
+                    require(swapAmounts[i] < 0, "Invalid amount of BPT");
+                    bptAmount = uint256(-swapAmounts[i]);
+                    break;
+                }
+            }
+
+            // Here we overwrite the bptAmountIn field of an `exactBptInForTokenOut` exit with the output of the swap
+            // We are mutating the userData field within request so no explicit assignment occurs
+            bytes memory userData = request.userData;
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                mstore(add(userData, _BPT_AMOUNT_IN_OFFSET), bptAmount)
+            }
+        }
+
+        getVault().exitPool(poolId, msg.sender, recipient, request);
+        _sweepETH();
+    }
+
+    function _sweepETH() private {
+        uint256 remainingEth = address(this).balance;
+        if (remainingEth > 0) {
+            msg.sender.sendValue(remainingEth);
+        }
+    }
```

Comment:
I'm worried this and other reliance on global state might open the door to reentrancy vulns - we might be fine since the Vault is non-reentrant, but I'd still consider adding the `nonReentrant` modifier to external fucntions.